tools/traffic_vault_migrate/postgres.go

```diff
@@ -0,0 +1,716 @@
+	}
+
+	pg.db = db
+	pg.sslKey = PGSSLKeyTable{}
+	pg.dnssec = PGDNSSecTable{}
+	pg.url = PGURLSigKeyTable{}
+	pg.uri = PGURISignKeyTable{}
+
+	return nil
+}
+func (pg *PGBackend) ValidateKey() []string {
+	var errors []string
+	if errs := pg.sslKey.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	if errs := pg.dnssec.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	if errs := pg.uri.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	if errs := pg.url.validate(); errs != nil {
+		errors = append(errors, errs...)
+	}
+	return errors
+}
+func (pg *PGBackend) Stop() error {
+	return pg.db.Close()
+}
+func (pg *PGBackend) Ping() error {
+	return pg.db.Ping()
+}
+func (pg *PGBackend) Fetch() error {
+	if err := pg.sslKey.gatherKeys(pg.db); err != nil {
+		return err
+	}
+
+	if err := pg.dnssec.gatherKeys(pg.db); err != nil {
+		return err
+	}
+
+	if err := pg.url.gatherKeys(pg.db); err != nil {
+		return err
+	}
+
+	if err := pg.uri.gatherKeys(pg.db); err != nil {
+		return err
+	}
+
+	return nil
+}
+func (pg *PGBackend) GetSSLKeys() ([]SSLKey, error) {
+	if err := pg.sslKey.decrypt(pg.cfg.AESKey); err != nil {
+		return nil, err
+	}
+	return pg.sslKey.toGeneric(), nil
+}
+func (pg *PGBackend) SetSSLKeys(keys []SSLKey) error {
+	pg.sslKey.fromGeneric(keys)
+	return pg.sslKey.encrypt(pg.cfg.AESKey)
+}
+func (pg *PGBackend) GetDNSSecKeys() ([]DNSSecKey, error) {
+	if err := pg.dnssec.decrypt(pg.cfg.AESKey); err != nil {
+		return nil, err
+	}
+	return pg.dnssec.toGeneric(), nil
+}
+func (pg *PGBackend) SetDNSSecKeys(keys []DNSSecKey) error {
+	pg.dnssec.fromGeneric(keys)
+	return pg.dnssec.encrypt(pg.cfg.AESKey)
+}
+func (pg *PGBackend) GetURISignKeys() ([]URISignKey, error) {
+	if err := pg.uri.decrypt(pg.cfg.AESKey); err != nil {
+		return nil, err
+	}
+	return pg.uri.toGeneric(), nil
+}
+func (pg *PGBackend) SetURISignKeys(keys []URISignKey) error {
+	pg.uri.fromGeneric(keys)
+	return pg.uri.encrypt(pg.cfg.AESKey)
+}
+func (pg *PGBackend) GetURLSigKeys() ([]URLSigKey, error) {
+	if err := pg.url.decrypt(pg.cfg.AESKey); err != nil {
+		return nil, err
+	}
+	return pg.url.toGeneric(), nil
+}
+func (pg *PGBackend) SetURLSigKeys(keys []URLSigKey) error {
+	pg.url.fromGeneric(keys)
+	return pg.url.encrypt(pg.cfg.AESKey)
+}
+
+type PGCommonRecord struct {
+	DataEncrypted []byte
+	CommonRecord
+}
+
+type PGDNSSecRecord struct {
+	Key tc.DNSSECKeysTrafficVault
+	CDN string
+	PGCommonRecord
+}
+type PGDNSSecTable struct {
+	Records []PGDNSSecRecord
+}
+
+func (tbl *PGDNSSecTable) gatherKeys(db *sql.DB) error {
+	sz, err := getSize(db, "dnssec")
+	if err != nil {
+		log.Println("PGDNSSec gatherKeys: unable to determine size of dnssec table")
+	}
+	tbl.Records = make([]PGDNSSecRecord, sz)
+
+	rows, err := db.Query("SELECT cdn, data from dnssec")
+	if err != nil {
+		return fmt.Errorf("PGDNSSec gatherKeys: unable to query: %v", err)
+	}
+	defer rows.Close()
```

Comment:
nit: this is done commonly in our codebase, but it is throwing away an error return value. There's a convenience method provided by `github.com/apache/trafficcontrol/lib/go-log/` to log the error using `log.Error` if an error occurs - [`log.Close`](https://pkg.go.dev/github.com/apache/trafficcontrol/lib/go-log#Close) (or `log.Closef` to use format strings). Some static analysis tools won't detect that as deferring a Close on a Closer resource, though.

e.g.
```go
defer log.Close(rows, "Closing PGDNSSEC rows from gathering keys")
```

Reply:
I just realized you're using the standard `log` package, not `github.com/apache/trafficcontrol/lib/go-log` - so what I said doesn't apply unless you wanna switch libraries.

Reply:
Switched to `go-log`